src/pip/_internal/resolution/legacy/resolver.py

```diff
@@ -536,7 +543,8 @@ def add_req(subreq: Requirement, extras_requested: Iterable[str]) -> None:
         with indent_log():
             # We add req_to_install before its dependencies, so that we
             # can refer to it when adding dependencies.
-            if not requirement_set.has_requirement(req_to_install.name):
+            name = req_to_install.name
+            if name is None or not requirement_set.has_requirement(name):
```

Comment:
A RequirementSet can only contain named requirements, so if a requirment does not have a name at this point, it should never be in the set.

Reply:
> A RequirementSet can only contain named requirements,

This isn't true as-is (<https://github.com/pypa/pip/blob/c20c7890729c5e14e3c56e2dd3d3737b84d1f458/src/pip/_internal/req/req_set.py#L41>) but, functionally, for this case, yes.

Again though, I'd prefer an assert here instead of tacking this on the `if` statement (these asserts serve to show us places where we can benefit from a redesign that removes the implicit conditionals around these packages).